source/Web/Client/WebSocket/Fido.Web.Client.WebSocket.SignalR.pas

```diff
@@ -0,0 +1,354 @@
+
+    FLock: IReadWriteSync;
+    FResultEventsMethodsMap: IDictionary<Int64, TOnMethodResultReceived>;
+    FMethodsResultsMap: IDictionary<Int64, TSignalRMethodResult>;
+  private
+    function NextCallId: Int64;
+    function Negotiate(const Host: string; const Secured: Boolean; const CustomHeaders: TStrings): INegotiationResponse;
+    procedure Connect(const Host: string; const Secured: Boolean; const NegotiationResponse: INegotiationResponse; const OnReceivedMessage: TWebSocketOnReceivedMessage; const OnError: TWebSocketOnError);
+  public
+    class operator Initialize(out Dest: TSignalR);
+
+    constructor Create(const Host: string; const Secured: Boolean; const OnReceivedMessage: TTSignalROnReceivedMessage; const WebSocketClient: IWebSocketClient; const OnError: TWebSocketOnError = nil); overload;
+    constructor Create(const Host: string; const Secured: Boolean; const CustomHeaders: TStrings; const OnReceivedMessage: TTSignalROnReceivedMessage; const WebSocketClient: IWebSocketClient; const OnError: TWebSocketOnError = nil); overload;
+
+    function Send<TArguments>(const Hub: string; const Method: string; const Arguments: TArguments; const OnMethodResultReceived: TOnMethodResultReceived): TSignalRMethodResult; overload;
+    function Send(const Hub: string; const Method: string; const Arguments: string; const OnMethodResultReceived: TOnMethodResultReceived): TSignalRMethodResult; overload;
+
+    procedure Start;
+    procedure Stop;
+    function Started: Boolean;
+  end;
+
+implementation
+
+{ TSignalR }
+
+constructor TSignalR.Create(
+  const Host: string;
+  const Secured: Boolean;
+  const CustomHeaders: TStrings;
+  const OnReceivedMessage: TTSignalROnReceivedMessage;
+  const WebSocketClient: IWebSocketClient;
+  const OnError: TWebSocketOnError);
+var
+  LLock: IReadWriteSync;
+  LResultEventsMap: IDictionary<Int64, TOnMethodResultReceived>;
+  LMethodsResultsMap: IDictionary<Int64, TSignalRMethodResult>;
+begin
+  Guard.CheckNotNull<TTSignalROnReceivedMessage>(OnReceivedMessage, 'OnReceivedMessage');
+
+  FHost := Host;
+  FSecured := Secured;
+  FCustomHeaders := TStringList.Create;
+  FCustomHeaders.Value.AddStrings(CustomHeaders);
+  FWebSocketClient := Utilities.CheckNotNullAndSet(WebSocketClient, 'WebSocketClient');
+  FOnError := OnError;
+
+  LLock := FLock;
+  LResultEventsMap := FResultEventsMethodsMap;
+  LMethodsResultsMap := FMethodsResultsMap;
+
+ {$REGION ' OnReceivedWebsocketMessage '}
+  FOnReceivedWebsocketMessage := procedure(const Message: string)
+    var
+      JsonObject: Shared<TJSONObject>;
+      TryProcessMethodsResults: TTryProcessJson;
+      TryReceivedData: TTryProcessJson;
+    begin
+      if Message = '{}' then
+        Exit;
+
+      JsonObject := TJSONValue.ParseJSONValue(Message) as TJSONObject;
+      if not Assigned(JsonObject.Value) then
+        Exit;
+
+     {$REGION ' Manage methods callbacks '}
+      TryProcessMethodsResults := function(const JsonObject: TJSONObject): Boolean
```

Comment:
You could move this one and the TryReceivedData function to a private function. It would make this assigned procedure a bit easier to read. But since it works as is, I think it's fine.